### DISCUSSION TITLE
Fundamentos do C# - Linguagens de programação com C# - Byte - Não compreendi bem o porquê de -127 a 128 com sinal

### DISCUSSION THREAD

**Discussion by Marcos-Vitor123** [balta-io/forum#423] [Tira Dúvidas] [answered]
### Não compreendi bem o porquê de -127 a 128 com sinal

1. Entendi bem que 1 byte equivale de 0 a 255 e obviamente é positivo. 
2. Mas, uma coisa que não compreendi bem foi sobre o byte com sinal negativo onde é de -127 a 128. 
3. Fiquei confuso com esse 128 positivo, que para mim dá a entender que ele invade o espaço das casas que fica entre 0 a 255. 
4. Entenderam? 
5. Só para melhorar o que penso: Imagine uma régua transversal onde 0 é o centro e os números a direita são positivos de 0 até 255 e que os números a esquerda é de -1 até -127. 
6. Tipo: parece que ele usa de 0 a 128 positivo que pertence na realidade aos campos de 0 a 255 ou seja que usa dos mesmo números que estão envolvidos de 0 a 255. 
7. Porque não poderia ser de -1 a -127. E lembrando que se levar por esse lado, não dá os 256 contando com o 0.
8.  Fiquei confuso somente nisso. Alguém pode esclarecer isso?

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @Marcos-Vitor123 como vai?
No caso, o byte vai de 0 a 255 (Total de 256), o sbyte vai de -128 a 127 (Total de 256).

O que talvez esteja te confundindo seja o 0 (Zero). Se você somar 128 (Possibilidades negativa) mais (127) terá o valor 255 (O zero conta para ambos os lados).

O valor 255 vem da possibilidade total da soma de 8 bits que formam 1 byte (1111 1111 em binário é 255 em decimal).
Desta forma, se mantiver apenas o 127, no fim teríamos 254, perdendo um bit. No caso do .NET, eles escolheram acrescentar esse bit no lado negativo, usando o -128.

Podem ter outras linguagens que seja ao contrário, ou mesmo que utilize uma faixa diferente.

Ficou mais claro?

**Reply by Marcos-Vitor123** [reply]
Boa tarde, mestre @andrebaltieri! 

Ficou claro em tese: 

1. Entendi que o intervalo de -128 a 127 é de 256 incluindo o zero ou seja a soma de 128 negativos + 0 + 127 positivos é 256 no total.
2. O que fiquei curioso, foi de o porquê no `sbyte` escolherem esse intervalo  de -128 a 127 ao invés de qualquer outro intervalo? Como por exemplo um intervalo de somente negativos abaixo de zero: de **-255** a **0** ou **0** a **-255** apenas. É nesse sentido que digo.
3. Estou pesquisando sobre isso também para compreender melhor. Do restante até agora entendi. É esse fato apenas que me despertou minha curiosidade. 
4. Caso eu consiga compreeder antes. Aviso aqui deixando esclarecido o que entendi.

**Reply by Marcos-Vitor123** [reply]
Bom, pesquisando aqui, acho que compreendi finalmente através do chat GPT- 4 do bing da Microsoft aliando a sua resposta acima.

Confirma ou acrescenta algo caso eu erre!

1. Entendi que se escolher qualquer outro intervalo que seja diferente desse de -128 a 127. Por algum motivo não vai conseguir representar todos os números positivos e negativos ao mesmo tempo.
2. Então por exemplo a cada bit que é alterado, representa um número tanto positivo quanto negativo no caso do intervalo do `sbyte` de -128 a 127.
3. Já no caso do `byte` no intervalo de 0 a 255. É somente para os números positivos e que precisam de todos esse **bytes** que mesmo alterando 1 ou mais bits a cada vez, só vai gerar um números positivos diferente um do outro. 

Bom, foi isso que entendi meio que por cima.
